src/databricks/sql/cloudfetch/downloader.py

```diff
@@ -19,85 +24,126 @@ def __init__(self, downloadable_result_settings, t_spark_arrow_result_link):
-            }
-            session.proxies.update(proxy)
+            self.is_link_expired = True
+            return
 
-            # ProxyAuthentication -> static enum BASIC and NONE
-            if self.settings.proxy_auth == "BASIC":
-                session.auth = requests.auth.HTTPBasicAuth(self.settings.proxy_uid, self.settings.proxy_pwd)
+        session = requests.Session()
+        session.timeout = self.settings.download_timeout
 
         try:
+            # Get the file via HTTP request
             response = session.get(self.result_link.fileLink)
-            self.http_code = response.status_code
 
-            if self.http_code != 200:
+            if not response.ok:
                 self.is_file_downloaded_successfully = False
-            else:
-                if self.settings.is_lz4_compressed:
-                    compressed_data = response.content
-                    uncompressed_data, bytes_read = lz4.frame.decompress(compressed_data, return_bytes_read=True)
-                    if bytes_read < len(compressed_data):
-                        d_context = lz4.frame.create_decompression_context()
-                        start = 0
-                        uncompressed_data = bytearray()
-                        while start < len(compressed_data):
-                            data, num_bytes, e = lz4.frame.decompress_chunk(d_context, compressed_data[start:])
-                            uncompressed_data += data
-                            start += num_bytes
-                    self.result_file = uncompressed_data
-                else:
-                    self.result_file = response.content
-                self.is_file_downloaded_successfully = len(self.result_file) == self.result_link.bytesNum
-        except:
+                return
+
+            # Save (and decompress if needed) the downloaded file
+            compressed_data = response.content
+            decompressed_data = (
+                ResultSetDownloadHandler.decompress_data(compressed_data)
+                if self.settings.is_lz4_compressed
+                else compressed_data
+            )
+            self.result_file = decompressed_data
+
+            # The size of the downloaded file should match the size specified from TSparkArrowResultLink
+            self.is_file_downloaded_successfully = (
+                len(self.result_file) == self.result_link.bytesNum
+            )
+        except Exception as e:
+            logger.error(e)
             self.is_file_downloaded_successfully = False
 
         finally:
-            session.close()
+            session and session.close()
+            # Awaken threads waiting for this to be true which signals the run is complete
             self.is_download_finished.set()
-            self.download_completion_semaphore.release()
+
+    def _reset(self):
+        # Reset download-related flags for every retry of run()
```

Comment:
Use a docstring so that IDEs can show this info on hover.

```suggestion
        """
        Reset download-related flags for every retry of run()
        """
```